src/devices/machine/r65c52.cpp

```diff
@@ -0,0 +1,987 @@
+	save_item(NAME(m_rxc));
+	save_item(NAME(m_rts));
+	save_item(NAME(m_dtr));
+
+	save_item(NAME(m_xtal));
+	save_item(NAME(m_divide));
+	save_item(NAME(m_cts));
+	save_item(NAME(m_dsr));
+	save_item(NAME(m_dcd));
+	save_item(NAME(m_rxd));
+
+	save_item(NAME(m_wordlength));
+	save_item(NAME(m_stoplength));
+	save_item(NAME(m_brk));
+	save_item(NAME(m_echo_mode));
+	save_item(NAME(m_parity));
+
+	save_item(NAME(m_rx_state));
+	save_item(NAME(m_rx_clock));
+	save_item(NAME(m_rx_bits));
+	save_item(NAME(m_rx_shift));
+	save_item(NAME(m_rx_parity));
+	save_item(NAME(m_rx_counter));
+
+	save_item(NAME(m_tx_state));
+	save_item(NAME(m_tx_output));
+	save_item(NAME(m_tx_clock));
+	save_item(NAME(m_tx_bits));
+	save_item(NAME(m_tx_shift));
+	save_item(NAME(m_tx_parity));
+
+	save_item(NAME(m_tx_counter));
+
+	m_internal_clock1->set_unscaled_clock(m_xtal);
+	m_internal_clock2->set_unscaled_clock(m_xtal);
+
+	set_xtal(m_xtal);
+}
+
+void r65c52_device::device_reset()
+{
+	for (int i = 0; i < 2; i++)
+	{
+		m_tdre[i] = true;
+		m_rdrf[i] = true;
+
+		m_status[i] = SR_FRAMING_ERROR;
+
+		m_aux_ctrl[i] = 0;
+		m_compare[i] = 0;
+		m_rdr[i] = 0x00;
+		m_isr[i] = 0x80;
+		m_ier[i] = 0x80;
+
+		m_rts[i] = 1;
+		m_dtr[i] = 1;
+
+		output_rts(i, 1);
+
+		output_dtr(i, 1);
+
+		m_rx_state[i] = STATE_START;
+		m_rx_counter[i] = 0;
+	}
+}
+
+u8 r65c52_device::stoplengthcounter(int idx)
+{
+	if (m_stoplength[idx] == 2)
+	{
+		if (m_wordlength[idx] == 5 && m_parity[idx] == PARITY_NONE)
+		{
+			return m_divide[idx] + (m_divide[idx] / 2);
+		}
+
+		if (m_wordlength[idx] < 8 || m_parity[idx] == PARITY_NONE)
+		{
+			return m_divide[idx] * 2;
+		}
+	}
+
+	return m_divide[idx];
+}
+
+void r65c52_device::output_irq(int idx, int irq)
+{
+	LOG("R65C52: %x  IRQ %x \n", idx + 1, irq);
+
+	if (m_irq[idx] != irq)
+	{
+		m_irq[idx] = irq;
+
+		m_irq_handler[idx](!m_irq[idx]);
+	}
+}
+
+void r65c52_device::output_txd(int idx, int txd)
+{
+	switch (m_tx_output[idx])
+	{
+	case OUTPUT_MARK:
+		txd = 1;
+		break;
+
+	case OUTPUT_BREAK:
+		txd = 0;
+		break;
+	}
+
+	if (m_txd[idx] != txd)
+	{
+		m_txd[idx] = txd;
+		m_txd_handler[idx](m_txd[idx]);
+	}
+}
+
+void r65c52_device::output_rts(int idx, int rts)
+{
+	LOG("R65C52: %x  RTS %x \n", idx + 1, rts);
+	if (m_rts[idx] != rts)
+	{
+		m_rts[idx] = rts;
+		m_rts_handler[idx](m_rts[idx]);
+		if (rts)
+		{
+			m_status[idx] |= SR_RTS;
+		}
+		else
+		{
+			m_status[idx] &= ~SR_RTS;
+		}
+	}
+}
+
+void r65c52_device::output_dtr(int idx, int dtr)
+{
+	LOG("R65C52: %x  DTR %x \n", idx + 1, dtr);
+	if (m_dtr[idx] != dtr)
+	{
+		m_dtr[idx] = dtr;
+		m_dtr_handler[idx](m_dtr[idx]);
+		if (dtr)
+		{
+			m_status[idx] |= SR_DTR;
+		}
+		else
+		{
+			m_status[idx] &= ~SR_DTR;
+		}
+	}
+}
+
+void r65c52_device::update_irq(int idx)
+{
+	bool irq = false;
+	LOG("R65C52: %x  IER  %x ISR %x\n", idx + 1, m_ier[idx], m_isr[idx]);
+	for (int i = 0; i < 8; i++)
+	{
+		if ((m_ier[idx] & (1 >> i)) && ((m_isr[idx] & (1 >> i))))
+		{
+			irq = true;
+		}
+	}
```

Comment:
How is this supposed to work?  `(1 >> i)` will be zero whenever `i` is not zero, so the loop is pointless – it only checks a bit on the first iteration.

Reply:
That's meant to be a bitwise comparison (i.e is the same bit set in the relevant slice of m_ier as m_isr). However, it only does that once, as you say. I have a fix, that I will submit separately, as this has already been merged.

Reply:
You know you don’t need a loop at all to test if bits on corresponding positions are set…

Reply:
Undoubtedly, but I'm still finding my feet with bitwise operations of this nature, so I thought I'd keep it relatively simple in the first phase rather than try an XOR. An optimisation might be useful here though, since this triggers a lot, but I'm not sure what best practice is. For now, I've been using the BIT macro to do it

Reply:
An XOR is wrong.  You want the intersection, not the disjunction.  And the loop is definitely not “simple” as this shows.